libs/datasource/src/datasource.cpp

```diff
@@ -0,0 +1,167 @@
+    {
+        // Set up GET /tile endpoint
+        server_.Get(
+            "/tile",
+            [this](const httplib::Request& req, httplib::Response& res)
+            {
+                auto layerIdParam = req.get_param_value("layer");
+                auto layerIt = info_.layers_.find(layerIdParam);
+                if (layerIt == info_.layers_.end())
+                    throw std::runtime_error(stx::format("Unknown layer id `{}`!", layerIdParam));
+
+                auto tileIdParam = TileId{std::stoull(req.get_param_value("tileId"))};
+                auto fieldsOffsetParam = (simfil::FieldId)0;
+                if (req.has_param("fieldsOffset"))
+                    fieldsOffsetParam = (simfil::FieldId)
+                        std::stoul(req.get_param_value("fieldsOffset"));
+
+                auto tileFeatureLayer = std::make_shared<TileFeatureLayer>(
+                    tileIdParam,
+                    info_.nodeId_,
+                    info_.mapId_,
+                    layerIt->second,
+                    fields_);
+                if (tileCallback_)
+                    tileCallback_(*tileFeatureLayer);
+
+                // Serialize TileFeatureLayer using TileLayerStream
+                std::stringstream content;
+                TileLayerStream::FieldOffsetMap fieldOffsets{{info_.nodeId_, fieldsOffsetParam}};
+                TileLayerStream::Writer layerWriter{
+                    [&](auto&& msg) { content << msg; },
+                    fieldOffsets};
+                layerWriter.write(tileFeatureLayer);
+
+                res.set_content(content.str(), "application/binary");
+            });
+
+        // Set up GET /info endpoint
+        server_.Get(
+            "/info",
+            [this](const httplib::Request&, httplib::Response& res)
+            {
+                nlohmann::json j = info_.toJson();
+                res.set_content(j.dump(), "application/json");
+            });
+
+    }
+
+    static void handleSignal(int signal)
+    {
+        // Stop the active instance when a signal is received.
+        if (activeDataSource != nullptr) {
+            (*activeDataSource).stop();
+            activeDataSource = nullptr;
+        }
+    }
+};
+
+DataSource::DataSource(DataSourceInfo const& info)
+    : impl_(new Impl(info))
+{
+}
+
+DataSource& DataSource::onTileRequest(std::function<void(TileFeatureLayer&)> const& callback)
+{
+    impl_->tileCallback_ = callback;
+    return *this;
+}
+
+void DataSource::go(std::string const& interfaceAddr, uint16_t port, uint32_t waitMs)
+{
+    if (impl_->server_.is_running() || impl_->serverThread_.joinable())
+        throw std::runtime_error("DataSource is already running");
+
+    if (port == 0) {
+        impl_->port_ = impl_->server_.bind_to_any_port(interfaceAddr);
+    }
+    else {
+        impl_->port_ = port;
+        impl_->server_.bind_to_port(interfaceAddr, port);
+    }
+
+    impl_->serverThread_ = std::thread(
```

Comment:
Wouldnt something like: `auto done = std::async(... return true)` plus `done.wait_for(...)` make more sense instead of always waiting a fixed amount of time?
Only affects failure case though.

Reply:
Hmm, I will stick with the current solution.